### PR TITLE
Do not copy up when volume is not empty

### PR DESCRIPTION
**Pull request by mheon**
When Docker performs a copy up, it first verifies that the volume being copied into is empty; thus, for volumes that have been modified elsewhere (e.g. manually copying into then), the copy up will not be performed at all. Duplicate this behavior in Podman by checking if the volume is empty before copying.

Furthermore, move setting copyup to false further up. This will prevent a potential race where copy up could happen more than once if Podman was killed after some files had been copied but before the DB was updated.

### CODE REVIEW

**Comment by mheon** [comment]
@vrothberg @rhatdan @baude @giuseppe @TomSweeneyRedHat PTAL

File: libpod/container_internal.go [hunk 1]
```diff
@@ -1383,18 +1383,34 @@ func (c *Container) mountNamedVolume(v *ContainerNamedVolume, mountpoint string)
...
+		// If the volume is not empty, we should not copy up.
+		volMount := vol.MountPoint()
+		contents, err := ioutil.ReadDir(volMount)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error listing contents of volume %s mountpoint when copying up from container %s", vol.Name(), c.ID())
+		}
+		if len(contents) > 0 {
+			// The volume is not empty. It was likely modified
+			// outside of Podman. For safety, let's not copy up into
+			// it.
```
Comment (vrothberg) [inline libpod/container_internal.go]: Maybe worth dropping the CVE in the comment (for historical reasons).

Comment (mheon) [reply libpod/container_internal.go]: Fair point

File: libpod/container_internal.go [hunk 2]
```diff
@@ -1383,18 +1383,34 @@ func (c *Container) mountNamedVolume(v *ContainerNamedVolume, mountpoint string)
...
+			return nil, err
+		}
+
+		// If the volume is not empty, we should not copy up.
+		volMount := vol.MountPoint()
+		contents, err := ioutil.ReadDir(volMount)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error listing contents of volume %s mountpoint when copying up from container %s", vol.Name(), c.ID())
+		}
+		if len(contents) > 0 {
```
Comment (rhatdan) [inline libpod/container_internal.go]: I take it contents does not include "." and ".."

Comment (mheon) [reply libpod/container_internal.go]: Yes, those are excluded

**Comment by TristanCacqueray** [comment]
I confirm dd65ad5a fixes the issue. Here is a reproducer:

```
# step1: create the image
CTX=$(buildah from registry.fedoraproject.org/fedora:31)
buildah run $CTX sh -c 'mkdir /data; echo oops > /data/secret.txt'
buildah commit --rm $CTX fedora-data

# step2: create the volume
podman volume create test
echo readonly-file > ~/.local/share/containers/storage/volumes/test/_data/secret.txt

# step3: trigger issue
podman run -it --rm --volume test:/data:ro fedora-data id
cat ~/.local/share/containers/storage/volumes/test/_data/secret.txt
```

The last stdout line should be `readonly-file`, not `oops`

**Comment by mheon** [comment]
Added a test

**Comment by vrothberg** [comment]
gating isn't happy yet

**Comment by mheon** [comment]
Tests seem to be coming back as flakes, not failures, this time. Might be ready with some retesting.

**Comment by mheon** [comment]
It's green. @rhatdan @baude @vrothberg @giuseppe @TomSweeneyRedHat PTAL and merge

**Comment by mheon** [comment]
Nevermind, need to fix one thing in tests

**Comment by mheon** [comment]
Re-pushed, should be good now. Held LGTM would be appreciated.

**Pull request merged**